File: BinarySearchTree/src/main/java/Solution450.java
import common.TreeNode;

public class Solution450 {
    public TreeNode deleteNode(TreeNode root, int key) {
        if (root == null) {
            return null;
        }

        if (root.val == key) {
            if (root.left == null) {
                // root = root.right;
                // 要直接返回， 否则，下面的可以能也进行了赋值
                return root.right;
            }
            if (root.right == null) {
                // root = root.left;
                return root.left;
            }

            TreeNode next = root.right;

            while (next.left != null) {
                next = next.left;
            }

            // 要赋值，否则会改变结构
            root.right = deleteNode(root.right, next.val);

            root.val = next.val;
            return root;
        }

        if (root.val < key) {
            root.right = deleteNode(root.right, key);
        } else {
            root.left = deleteNode(root.left, key);
        }

        return root;
    }
}

File: BinarySearchTree/src/main/python/solution450.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if not root:
            return

        if root.val == key:
            if root.left == None:
                return root.right
            if root.right == None:
                return root.left

            next = root.right

            while next.left:
                next = next.left

            root.right = self.deleteNode(root.right, next.val)
            root.val = next.val

            return root

        if root.val < key:
            root.right = self.deleteNode(root.right, key)

        else:
            root.left = self.deleteNode(root.left, key)

        return root
